File: src/data/telemetry_lap_data.cpp
#include "data/telemetry_lap_data.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Data
{
VectorDouble::VectorDouble(const PbData::VectorDouble& protobuf) {
    buffer = {protobuf.buffer().begin(), protobuf.buffer().end()};
}

VectorDouble::operator PbData::VectorDouble() const {
    PbData::VectorDouble ret;
    *(ret.mutable_buffer()) = {buffer.begin(), buffer.end()};
    return ret;
}

std::string VectorDouble::serializeAsJsonString() const {
    PbData::VectorDouble protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string VectorDouble::serializeAsProtobufString() const {
    PbData::VectorDouble protobuf(*this);
    return protobuf.SerializeAsString();
}

bool VectorDouble::deserializeFromJsonString(const std::string& str) {
    PbData::VectorDouble protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool VectorDouble::deserializeFromProtobufString(const std::string& str) {
    PbData::VectorDouble protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

DataFile::DataFile(const PbData::DataFile& protobuf) {
    data = {protobuf.data().begin(), protobuf.data().end()};
}

DataFile::operator PbData::DataFile() const {
    PbData::DataFile ret;
    *(ret.mutable_data()) = {data.begin(), data.end()};
    return ret;
}

std::string DataFile::serializeAsJsonString() const {
    PbData::DataFile protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string DataFile::serializeAsProtobufString() const {
    PbData::DataFile protobuf(*this);
    return protobuf.SerializeAsString();
}

bool DataFile::deserializeFromJsonString(const std::string& str) {
    PbData::DataFile protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool DataFile::deserializeFromProtobufString(const std::string& str) {
    PbData::DataFile protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

TelemetryLapData::TelemetryLapData(const PbData::TelemetryLapData& protobuf) {
    dateTime = protobuf.datetime();
    trackLocation = protobuf.tracklocation();
    trackLayout = protobuf.tracklayout();
    driver = protobuf.driver();
    lapNumber = protobuf.lapnumber();
    filenameHash = protobuf.filenamehash();
}

TelemetryLapData::operator PbData::TelemetryLapData() const {
    PbData::TelemetryLapData ret;
    ret.set_datetime(dateTime);
    ret.set_tracklocation(trackLocation);
    ret.set_tracklayout(trackLayout);
    ret.set_driver(driver);
    ret.set_lapnumber(lapNumber);
    ret.set_filenamehash(filenameHash);
    return ret;
}

std::string TelemetryLapData::serializeAsJsonString() const {
    PbData::TelemetryLapData protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string TelemetryLapData::serializeAsProtobufString() const {
    PbData::TelemetryLapData protobuf(*this);
    return protobuf.SerializeAsString();
}

bool TelemetryLapData::deserializeFromJsonString(const std::string& str) {
    PbData::TelemetryLapData protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool TelemetryLapData::deserializeFromProtobufString(const std::string& str) {
    PbData::TelemetryLapData protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

DataBase::DataBase(const PbData::DataBase& protobuf) {
    lapsData = {protobuf.lapsdata().begin(), protobuf.lapsdata().end()};
}

DataBase::operator PbData::DataBase() const {
    PbData::DataBase ret;
    *(ret.mutable_lapsdata()) = {lapsData.begin(), lapsData.end()};
    return ret;
}

std::string DataBase::serializeAsJsonString() const {
    PbData::DataBase protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string DataBase::serializeAsProtobufString() const {
    PbData::DataBase protobuf(*this);
    return protobuf.SerializeAsString();
}

bool DataBase::deserializeFromJsonString(const std::string& str) {
    PbData::DataBase protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool DataBase::deserializeFromProtobufString(const std::string& str) {
    PbData::DataBase protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}

File: py/data/telemetry_lap_data.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from data import telemetry_lap_data_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class VectorDouble:
    buffer: List[float] = field(default_factory=list)
    
    _proto_message: telemetry_lap_data_pb2.VectorDouble = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = telemetry_lap_data_pb2.VectorDouble()

    def _populate_proto(self):
        del self._proto_message.buffer[:]
        for val in self.buffer:
            self._proto_message.buffer.append(val)

    @classmethod
    def _from_proto(cls, proto_message) -> "VectorDouble":
        return cls(
            buffer=[float(val) for val in proto_message.buffer],
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "VectorDouble":
        message = telemetry_lap_data_pb2.VectorDouble()
        message.ParseFromString(data)
        return cls(
            buffer = [float(val) for val in message.buffer],
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "VectorDouble":
        message = telemetry_lap_data_pb2.VectorDouble()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class DataFile:
    data: Dict[str, VectorDouble] = field(default_factory=dict)
    
    _proto_message: telemetry_lap_data_pb2.DataFile = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = telemetry_lap_data_pb2.DataFile()

    def _populate_proto(self):
        self._proto_message.data.clear()
        for key, val in self.data.items():
            val._populate_proto()
            tmp = self._proto_message.data.setdefault(key)
            tmp.CopyFrom(val._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "DataFile":
        return cls(
            data={key: VectorDouble._from_proto(val) for key, val in proto_message.data.items()},
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "DataFile":
        message = telemetry_lap_data_pb2.DataFile()
        message.ParseFromString(data)
        return cls(
            data = {key: VectorDouble._from_proto(val) for key, val in message.data.items()},
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "DataFile":
        message = telemetry_lap_data_pb2.DataFile()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class TelemetryLapData:
    dateTime: str = ""
    trackLocation: str = ""
    trackLayout: str = ""
    driver: str = ""
    lapNumber: int = 0
    filenameHash: str = ""
    
    _proto_message: telemetry_lap_data_pb2.TelemetryLapData = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = telemetry_lap_data_pb2.TelemetryLapData()

    def _populate_proto(self):
        self._proto_message.dateTime = self.dateTime
        self._proto_message.trackLocation = self.trackLocation
        self._proto_message.trackLayout = self.trackLayout
        self._proto_message.driver = self.driver
        self._proto_message.lapNumber = self.lapNumber
        self._proto_message.filenameHash = self.filenameHash

    @classmethod
    def _from_proto(cls, proto_message) -> "TelemetryLapData":
        return cls(
            dateTime = proto_message.dateTime,
            trackLocation = proto_message.trackLocation,
            trackLayout = proto_message.trackLayout,
            driver = proto_message.driver,
            lapNumber = proto_message.lapNumber,
            filenameHash = proto_message.filenameHash,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "TelemetryLapData":
        message = telemetry_lap_data_pb2.TelemetryLapData()
        message.ParseFromString(data)
        return cls(
            dateTime = message.dateTime,
            trackLocation = message.trackLocation,
            trackLayout = message.trackLayout,
            driver = message.driver,
            lapNumber = message.lapNumber,
            filenameHash = message.filenameHash,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "TelemetryLapData":
        message = telemetry_lap_data_pb2.TelemetryLapData()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class DataBase:
    lapsData: List[TelemetryLapData] = field(default_factory=list)
    
    _proto_message: telemetry_lap_data_pb2.DataBase = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = telemetry_lap_data_pb2.DataBase()

    def _populate_proto(self):
        del self._proto_message.lapsData[:]
        for val in self.lapsData:
            val._populate_proto()
            tmp = self._proto_message.lapsData.add()
            tmp.CopyFrom(val._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "DataBase":
        return cls(
            lapsData=[TelemetryLapData._from_proto(val) for val in proto_message.lapsData],
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "DataBase":
        message = telemetry_lap_data_pb2.DataBase()
        message.ParseFromString(data)
        return cls(
            lapsData = [TelemetryLapData._from_proto(val) for val in message.lapsData],
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "DataBase":
        message = telemetry_lap_data_pb2.DataBase()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
